### ISSUE TITLE
AWS Sagemaker (SNMG & MNMG)

### ISSUE THREAD

**Issue by mmccarty** [rapidsai/deployment#13]

**Comment by skirui-source** [comment]
Experiencing the following blockers: 

1. The notebook examples in the provided [repo](https://github.com/shashankprasanna/sagemaker-rapids) is 2+ years old;  the cuml and cudf API have drastically changed ever since. For instance, some of the hyperparameteres for cuml` RandomForestClassifer` have been deprecated -- excluded them

2. ` sagemaker-containers` (package that allows rapids container to be compatible with SageMaker) has also been[ deprecated ](https://github.com/aws/sagemaker-containers/issues/274#issuecomment-658892088), and is replaced with [`sagemaker-training`](https://github.com/aws/sagemaker-training-toolkit) and [`sagemaker-inference` ](https://github.com/aws/sagemaker-inference-toolkit) toolkits

3. Lots of dependency conflicts btw  the latest rapids release v22.10 and sagemaker-training v3.4  (conflicting version requirements for some packages: cuda-python, cupy-cuda, protobuf, botocore)

**Comment by skirui-source** [comment]
Seeing this error with `rapids_estimator.fit({'dataset': s3_data_dir})`, blocking a completed training job:

```
I 2022-11-16 03:05:01.454 ServerApp] jupyterlab | extension was successfully loaded.
[W 2022-11-16 03:05:01.454 ServerApp] jupyterlab_nvdashboard | extension failed loading with message: 'NoneType' object is not callable
[E 2022-11-16 03:05:01.454 ServerApp] jupyterlab_nvdashboard | stack trace
    Traceback (most recent call last):
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 355, in load_extension
        extension.load_all_points(self.serverapp)
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 229, in load_all_points
        return [self.load_point(point_name, serverapp) for point_name in self.extension_points]
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 229, in <listcomp>
        return [self.load_point(point_name, serverapp) for point_name in self.extension_points]
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py"
```, line 222, in load_point
        return point.load(serverapp)
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 148, in load
        return loader(serverapp)
    TypeError: 'NoneType' object is not callable

**Comment by jacobtomlinson** [comment]
Hrm that error is related to nvdashboard which I don't think is relevant to the `.fit` call. Is there any other output or logs?

**Comment by skirui-source** [comment]
Actually I don't think [this](https://github.com/rapidsai/deployment/issues/13#issuecomment-1316289087) error affects training (heres the full output): 

```
2022-11-21 20:40:43 Starting - Starting the training job...
2022-11-21 20:41:08 Starting - Preparing the instances for trainingProfilerReport-1669063243: InProgress
.....................
2022-11-21 20:44:29 Downloading - Downloading input data...........[WARN  tini (6)] Tini is not running as PID 1 .
Zombie processes will not be re-parented to Tini, so zombie reaping won't work.
To fix the problem, run Tini as PID 1.
This container image and its contents are governed by the NVIDIA Deep Learning Container License.
By pulling and using the container, you accept the terms and conditions of this license:
https://developer.download.nvidia.com/licenses/NVIDIA_Deep_Learning_Container_License.pdf
[I 2022-11-21 20:46:27.832 ServerApp] dask_labextension | extension was successfully linked.
[I 2022-11-21 20:46:27.833 ServerApp] jupyter_server_proxy | extension was successfully linked.
[W 2022-11-21 20:46:27.837 LabApp] 'token' has moved from NotebookApp to ServerApp. This config will be passed to ServerApp. Be sure to update your config before our next release.
[W 2022-11-21 20:46:27.837 LabApp] 'allow_origin' has moved from NotebookApp to ServerApp. This config will be passed to ServerApp. Be sure to update your config before our next release.
[W 2022-11-21 20:46:27.837 LabApp] 'base_url' has moved from NotebookApp to ServerApp. This config will be passed to ServerApp. Be sure to update your config before our next release.
[I 2022-11-21 20:46:27.844 ServerApp] jupyterlab | extension was successfully linked.
[I 2022-11-21 20:46:27.844 ServerApp] jupyterlab_nvdashboard | extension was successfully linked.
[I 2022-11-21 20:46:27.854 ServerApp] nbclassic | extension was successfully linked.
[I 2022-11-21 20:46:27.864 ServerApp] Writing Jupyter server cookie secret to /root/.local/share/jupyter/runtime/jupyter_cookie_secret
[W 2022-11-21 20:46:27.874 ServerApp] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
[I 2022-11-21 20:46:27.876 ServerApp] dask_labextension | extension was successfully loaded.
[I 2022-11-21 20:46:28.353 ServerApp] jupyter_server_proxy | extension was successfully loaded.
[I 2022-11-21 20:46:28.354 LabApp] JupyterLab extension loaded from /opt/conda/envs/rapids/lib/python3.9/site-packages/jupyterlab
[I 2022-11-21 20:46:28.354 LabApp] JupyterLab application directory is /opt/conda/envs/rapids/share/jupyter/lab
[I 2022-11-21 20:46:28.357 ServerApp] jupyterlab | extension was successfully loaded.
[W 2022-11-21 20:46:28.357 ServerApp] jupyterlab_nvdashboard | extension failed loading with message: 'NoneType' object is not callable
[E 2022-11-21 20:46:28.357 ServerApp] jupyterlab_nvdashboard | stack trace
    Traceback (most recent call last):
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 355, in load_extension
        extension.load_all_points(self.serverapp)
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 229, in load_all_points
        return [self.load_point(point_name, serverapp) for point_name in self.extension_points]
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 229, in <listcomp>
        return [self.load_point(point_name, serverapp) for point_name in self.extension_points]
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 222, in load_point
        return point.load(serverapp)
      File "/opt/conda/envs/rapids/lib/python3.9/site-packages/jupyter_server/extension/manager.py", line 148, in load
        return loader(serverapp)
    TypeError: 'NoneType' object is not callable
[I 2022-11-21 20:46:28.361 ServerApp] nbclassic | extension was successfully loaded.
[I 2022-11-21 20:46:28.361 ServerApp] Serving notebooks from local directory: /rapids/notebooks
[I 2022-11-21 20:46:28.361 ServerApp] Jupyter Server 1.21.0 is running at:
[I 2022-11-21 20:46:28.361 ServerApp] http://ip-10-0-119-52.us-west-1.compute.internal:8888/lab
[I 2022-11-21 20:46:28.362 ServerApp]  or http://127.0.0.1:8888/lab
[I 2022-11-21 20:46:28.362 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).

2022-11-21 20:46:29 Training - Training image download completed. Training in progress.
​

```

I'm just not sure how long training should be, it's been an hour and still no output?

**Comment by skirui-source** [comment]
The training job fails with this output: 
```
2022-11-22 06:37:45 Starting - Starting the training job...
2022-11-22 06:38:08 Starting - Preparing the instances for trainingProfilerReport-1669099065: InProgress
...............
2022-11-22 06:40:41 Downloading - Downloading input data....................Running SageMaker HPO entrypoint.
@ entrypoint -> launching training script 
2022-11-22 06:43:58,978     INFO hpo_log 
parsing configuration from environment settings...
2022-11-22 06:43:58,978     INFO hpo_log   Dataset: Airline
  Compute: single-GPU
  Algorithm: XGBoost
  CV_folds: 3
2022-11-22 06:43:58,978     INFO hpo_log parsing model hyperparameters from command line arguments...log
2022-11-22 06:43:58,979     INFO hpo_log {    'gamma': 0.0,
     'lambda': 1,
     'learning_rate': 0.3,
     'max_depth': 5,
     'num_boost_round': 10,
     'objective': 'binary:logistic',
     'random_state': 0,
     'seed': 0,
     'tree_method': 'gpu_hist',
     'verbosity': 0}
2022-11-22 06:43:58,980     INFO hpo_log CSV input files detected
['/opt/ml/input/data/training/HIGGS.csv']
2022-11-22 06:43:58,980     INFO hpo_log detected 1 files as input
Traceback (most recent call last):
  File "cuda/_cuda/ccuda.pyx", line 3671, in cuda._cuda.ccuda._cuInit
  File "cuda/_cuda/ccuda.pyx", line 435, in cuda._cuda.ccuda.cuPythonInit
RuntimeError: Failed to dlopen libcuda.so
Exception ignored in: 'cuda._lib.ccudart.utils.cudaPythonGlobal.lazyInitGlobal'
Traceback (most recent call last):
  File "cuda/_cuda/ccuda.pyx", line 3671, in cuda._cuda.ccuda._cuInit
  File "cuda/_cuda/ccuda.pyx", line 435, in cuda._cuda.ccuda.cuPythonInit
RuntimeError: Failed to dlopen libcuda.so
/opt/rapids_cloudml/entrypoint.sh: line 45:    15 Segmentation fault      (core dumped) python train.py

2022-11-22 06:44:14 Training - Training image download completed. Training in progress.
2022-11-22 06:44:14 Uploading - Uploading generated training model
2022-11-22 06:44:14 Failed - Training job failed
```

**Comment by skirui-source** [comment]
^^ Here is the complete Traceback:

**Comment by jacobtomlinson** [comment]
The `RuntimeError: Failed to dlopen libcuda.so` error suggests that the VMs the estimator spins up have some driver or container runtime issue. You might want to check in with the cuml folks in case they have some thoughts.

**Comment by skirui-source** [comment]
Seems like the segfault error is due to outdated version of `cuda-python` dependency, following this [PR] merge (https://github.com/rapidsai/cuml/pull/4961) ,  on hold - waiting on RAPIDS v22.12 to be pushed to Docker Hub

**Comment by jacobtomlinson** [comment]
Are you able to test with the current nightly?

**Comment by skirui-source** [comment]
Update: now seeing this error  when I run  `estimator.fit(inputs = s3_data_input, job_name = job_name.lower())`
- training on rapids v22.12 stable that was released today:

```
estimator.fit(inputs = s3_data_input, job_name = job_name.lower())

2022-12-13 23:08:17 Starting - Starting the training job...ProfilerReport-1670972896: InProgress
...
2022-12-13 23:09:08 Starting - Preparing the instances for training......
2022-12-13 23:10:11 Downloading - Downloading input data...
2022-12-13 23:10:48 Training - Downloading the training image.....................
2022-12-13 23:14:19 Uploading - Uploading generated training model.@ entrypoint -> launching training script 
2022-12-13 23:14:13,723     INFO hpo_log 
parsing configuration from environment settings...
2022-12-13 23:14:13,723     INFO hpo_log   Dataset: Airline
  Compute: single-GPU
  Algorithm: XGBoost
  CV_folds: 3
2022-12-13 23:14:13,723     INFO hpo_log parsing model hyperparameters from command line arguments...log
2022-12-13 23:14:13,728     INFO hpo_log {    'gamma': 0.0,
     'lambda': 1,
     'learning_rate': 0.3,
     'max_depth': 5,
     'num_boost_round': 10,
     'objective': 'binary:logistic',
     'random_state': 0,
     'seed': 0,
     'tree_method': 'gpu_hist',
     'verbosity': 0}
2022-12-13 23:14:13,729     INFO hpo_log Parquet input files detected
['/opt/ml/input/data/training/part.6.parquet',
 '/opt/ml/input/data/training/part.1.parquet',
 '/opt/ml/input/data/training/part.12.parquet',
 '/opt/ml/input/data/training/part.10.parquet',
 '/opt/ml/input/data/training/part.2.parquet',
 '/opt/ml/input/data/training/part.8.parquet',
 '/opt/ml/input/data/training/part.11.parquet',
 '/opt/ml/input/data/training/part.0.parquet',
 '/opt/ml/input/data/training/part.7.parquet',
2022-12-13 23:14:13,729     INFO hpo_log detected 13 files as input
 '/opt/ml/input/data/training/part.3.parquet',
 '/opt/ml/input/data/training/part.5.parquet',
 '/opt/ml/input/data/training/part.4.parquet',
 '/opt/ml/input/data/training/part.9.parquet']
Traceback (most recent call last):
  File "/opt/conda/envs/rapids/lib/python3.9/site-packages/cupy/__init__.py", line 18, in <module>
    from cupy import _core  # NOQA
  File "/opt/conda/envs/rapids/lib/python3.9/site-packages/cupy/_core/__init__.py", line 3, in <module>
    from cupy._core import core  # NOQA
  File "cupy/_core/core.pyx", line 1, in init cupy._core.core
  File "/opt/conda/envs/rapids/lib/python3.9/site-packages/cupy/cuda/__init__.py", line 8, in <module>
    from cupy.cuda import compiler  # NOQA
  File "/opt/conda/envs/rapids/lib/python3.9/site-packages/cupy/cuda/compiler.py", line 13, in <module>
    from cupy.cuda import device
  File "cupy/cuda/device.pyx", line 5, in init cupy.cuda.device
ImportError: cannot import name syncdetect
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/opt/ml/code/train.py", line 76, in <module>
    train()
  File "/opt/ml/code/train.py", line 27, in train
    ml_workflow = create_workflow(hpo_config)
  File "/opt/ml/code/MLWorkflow.py", line 37, in create_workflow
    from workflows.MLWorkflowSingleGPU import MLWorkflowSingleGPU
  File "/opt/ml/code/workflows/MLWorkflowSingleGPU.py", line 21, in <module>
    import cudf
  File "/opt/conda/envs/rapids/lib/python3.9/site-packages/cudf/__init__.py", line 7, in <module>
    import cupy
  File "/opt/conda/envs/rapids/lib/python3.9/site-packages/cupy/__init__.py", line 20, in <module>
    raise ImportError(f'''
ImportError: 
================================================================
Failed to import CuPy.
If you installed CuPy via wheels (cupy-cudaXXX or cupy-rocm-X-X), make sure that the package matches with the version of CUDA or ROCm installed.
On Linux, you may need to set LD_LIBRARY_PATH environment variable depending on how you installed CUDA/ROCm.
On Windows, try setting CUDA_PATH environment variable.
Check the Installation Guide for details:
  https://docs.cupy.dev/en/latest/install.html
Original error:
  ImportError: cannot import name syncdetect
================================================================

2022-12-13 23:14:49 Failed - Training job failed
ProfilerReport-1670972896: NoIssuesFound
---------------------------------------------------------------------------
UnexpectedStatusException                 Traceback (most recent call last)
/tmp/ipykernel_5183/3160662281.py in <cell line: 1>()
----> 1 estimator.fit(inputs = s3_data_input, job_name = job_name.lower())

~/anaconda3/envs/python3/lib/python3.8/site-packages/sagemaker/workflow/pipeline_context.py in wrapper(*args, **kwargs)
    270             return _StepArguments(retrieve_caller_name(self_instance), run_func, *args, **kwargs)
    271 
--> 272         return run_func(*args, **kwargs)
    273 
    274     return wrapper

~/anaconda3/envs/python3/lib/python3.8/site-packages/sagemaker/estimator.py in fit(self, inputs, wait, logs, job_name, experiment_config)
   1126         self.jobs.append(self.latest_training_job)
   1127         if wait:
-> 1128             self.latest_training_job.wait(logs=logs)
   1129 
   1130     def _compilation_job_name(self):

~/anaconda3/envs/python3/lib/python3.8/site-packages/sagemaker/estimator.py in wait(self, logs)
   2240         # If logs are requested, call logs_for_jobs.
   2241         if logs != "None":
-> 2242             self.sagemaker_session.logs_for_job(self.job_name, wait=True, log_type=logs)
   2243         else:
   2244             self.sagemaker_session.wait_for_job(self.job_name)

~/anaconda3/envs/python3/lib/python3.8/site-packages/sagemaker/session.py in logs_for_job(self, job_name, wait, poll, log_type)
   4070 
   4071         if wait:
-> 4072             self._check_job_status(job_name, description, "TrainingJobStatus")
   4073             if dot:
   4074                 print()

~/anaconda3/envs/python3/lib/python3.8/site-packages/sagemaker/session.py in _check_job_status(self, job, desc, status_key_name)
   3601                     actual_status=status,
   3602                 )
-> 3603             raise exceptions.UnexpectedStatusException(
   3604                 message=message,
   3605                 allowed_statuses=["Completed", "Stopped"],

UnexpectedStatusException: Error for Training job air-sgpu-xgb-3cv-7811ba8b501f5b1: Failed. Reason: AlgorithmError: , exit code: 255
```

**Comment by jacobtomlinson** [comment]
I see ` ImportError: cannot import name syncdetect` is in the middle of the traceback. Perhaps that package is missing from the docker container?

**Comment by skirui-source** [comment]
4. `Multi Node, Multi GPU`
     - large HPO: 10_year, XGBoost, 5 CV folds, multiGPU, max_jobs = 10, max_parallel_jobs = 2 Random Search
     - tested on **2** `ml.p3.8xlarge` instances
